# Desafio Técnico Builders - Data Analytics
## Autor: Paulo França

### Etapas do projeto

1. Configuração do ambiente
2. Import de biblioteca
3. Extração dos dados
    - Dados mysql covid
    - Dados mongdb multas
4. Transformação de dados
    - Dados covid e mongdb normatizados
    - Criandos tabelas fato e dimeção das bases
5. Carregamento dos dados
    - Importando dados csv para leitura no Power BI    

### Configurando o ambiente

In [27]:
### Esses foram os pacotes necessários

#pip install pymysql
#pip install mysqlclient
#pip install pymongo
#pip install pandas

### Importando pacotes

In [28]:
import pandas as pd
from sqlalchemy import create_engine
from pymongo import MongoClient
from datetime import datetime, timedelta
import calendar

### Conexão e extração MySQL

In [29]:
# Configurações do banco de dados MySQL - Builders
db_config = {
    'host': '34.95.170.227',
    'user': 'teste-dados-leitura',
    'password': 'o7c4Cc8NDeXYbAMH',
    'database': 'teste_dados'
}

In [30]:
# Conectar ao banco de dados MySQL

db_uri = f"mysql://{db_config['user']}:{db_config['password']}@{db_config['host']}/{db_config['database']}"

# Conectar ao banco de dados usando SQLAlchemy
engine = create_engine(db_uri)    

query = "SELECT * FROM teste_dados.DADOS_COVID"

In [31]:
# Executar a consulta e carregar os dados em um DataFrame do pandas
df = pd.read_sql(query, engine)

In [32]:
display(df)

city  city_ibge_code        date  \
0                  Rio Branco         1200401  2020-03-17   
1                                          12  2020-03-17   
2                  Rio Branco         1200401  2020-03-18   
3                                          12  2020-03-18   
4                  Rio Branco         1200401  2020-03-19   
...                       ...             ...         ...   
13283  Porto Real do Colégio         2707503  2020-03-31   
13284                                      27  2020-03-31   
13285                 Maceió         2704302  2020-04-01   
13286  Porto Real do Colégio         2707503  2020-04-01   
13287                                      27  2020-04-01   

       epidemiological_week  estimated_population  estimated_population_2019  \
0                    202012                413418                     407319   
1                    202012                894470                     881935   
2                    202012                413418                     407319   
3                    202012                894470                     881935   
4                    202012                413418                     407319   
...                     ...                   ...                        ...   
13283                202014                 20112                      20066   
13284                202014               3351543                    3337357   
13285                202014               1025360                    1018948   
13286                202014                 20112                      20066   
13287                202014               3351543                    3337357   

      is_last is_repeated  last_available_confirmed  \
0       False       False                         3   
1       False       False                         3   
2       False       False                         3   
3       False       False                         3   
4       False       False                         4   
...       ...         ...                       ...   
13283   False       False                         1   
13284   False       False                        18   
13285   False       False                        13   
13286   False       False                         1   
13287   False       False                        18   

       last_available_confirmed_per_100k_inhabitants last_available_date  \
0                                            0.72566          2020-03-17   
1                                            0.33539          2020-03-17   
2                                            0.72566          2020-03-18   
3                                            0.33539          2020-03-18   
4                                            0.96754          2020-03-19   
...                                              ...                 ...   
13283                                        4.97216          2020-03-31   
13284                                        0.53707          2020-03-31   
13285                                        1.26785          2020-04-01   
13286                                        4.97216          2020-04-01   
13287                                        0.53707          2020-04-01   

       last_available_death_rate  last_available_deaths  order_for_place  \
0                         0.0000                      0                1   
1                         0.0000                      0                1   
2                         0.0000                      0                2   
3                         0.0000                      0                2   
4                         0.0000                      0                3   
...                          ...                    ...              ...   
13283                     0.0000                      0                7   
13284                     0.0556                      1               24   
13285                     0.0769                      1               25 

### Conexão e extração Mongdb

In [33]:
# String de conexão de desavio
string_de_conexao = "mongodb+srv://teste_dados_leitura:o7c4Cc8NDeXYbAMH@mongodbtestebuilders.vuzqjs5.mongodb.net/?retryWrites=true&w=majority"


In [34]:

# Conectando ao servidor MongoDB 
client = MongoClient(string_de_conexao)

# Acessando o banco de dados desejado 
db = client.teste_dados

#Checando coloções disponíveis
colecoes = db.list_collection_names()

In [35]:
### Aqui identificamos que existe apenas uma coleção no banco teste_dados
print("Coleções disponíveis:")
for colecao in colecoes:
    print(colecao)

Coleções disponíveis:
multas


In [36]:
# Acessando a coleção 
collection = db.multas

# Obtendo todos os documentos na coleção
result = collection.find()

In [37]:
# Convertendo os dados em um DataFrame do Pandas
df_multas = pd.DataFrame(list(result))

#  Print para verificar o dataframe convertido
print(df_multas)

                          _id  escopo_autuacao        mes   ano  uf  \
0    6407641b305a576376e6f372  Excesso de Peso  FEVEREIRO  2021  RJ   
1    6407641b305a576376e6f374  Excesso de Peso  FEVEREIRO  2021  RS   
2    6407641b305a576376e6f38f  Excesso de Peso      MARÇO  2021  RJ   
3    6407641b305a576376e6f3aa  Excesso de Peso      ABRIL  2021  ES   
4    6407641b305a576376e6f3b7  Excesso de Peso      ABRIL  2021  RJ   
..                        ...              ...        ...   ...  ..   
965  6407641b305a576376e6f6bf  Excesso de Peso   NOVEMBRO  2022  ES   
966  6407641b305a576376e6f6c0  Excesso de Peso   NOVEMBRO  2022  GO   
967  6407641b305a576376e6f6c1  Excesso de Peso   NOVEMBRO  2022  GO   
968  6407641b305a576376e6f6d4  Excesso de Peso   NOVEMBRO  2022  SC   
969  6407641b305a576376e6f6ee  Excesso de Peso   DEZEMBRO  2022  GO   

                                          amparo_legal  \
0    Lei 9503/97 Lei 10561/02 Res. CONTRAN nº 210/0...   
1    Lei 9503/97 Lei 10561/02 R

### Analise exploratória

##### Explorando as bases de dados

- A base de dados de covid possui dados de quantidade de mortes de cidade e estado em linhas distintas. É importante verificar se a soma dos dados de cidade representam ao do estado ou se devemos trata-las como tabelas fatos distintas.

- Com isso vou checar se a soma das ocorrências de morte de cidade e estado é a mesma.

In [38]:
### Pegando um exemplo do estado do Acre para avaliar se existe 

df[(df['place_type'] == 'city') & (df['state'] == 'AL') & (df['date'] == '2020-03-17')  ][['estimated_population','estimated_population_2019','new_confirmed','new_deaths']].agg(sum)

estimated_population         1025360
estimated_population_2019    1018948
new_confirmed                      0
new_deaths                         0
dtype: int64

In [39]:
### Aqui trago a soma da população por dia na visão cidade

df_res_city = df[(df['place_type'] == 'city')].groupby('date')[['estimated_population','estimated_population_2019','new_confirmed','new_deaths']].agg(sum)
print(df_res_city)


            estimated_population  estimated_population_2019  new_confirmed  \
date                                                                         
2020-03-08               1025360                    1018948              1   
2020-03-09               1025360                    1018948              0   
2020-03-10               1025360                    1018948              0   
2020-03-11               1025360                    1018948              0   
2020-03-12               1025360                    1018948              0   
...                          ...                        ...            ...   
2021-11-18                894470                     881935              0   
2021-11-19                894470                     881935              0   
2021-11-20                894470                     881935              0   
2021-11-21                894470                     881935              0   
2021-11-22                894470                     881935     

In [40]:
### Aqui trago a soma da população por dia na visão estado

df_res_state = df[(df['place_type'] == 'state')].groupby('date')[['estimated_population','estimated_population_2019','new_confirmed','new_deaths']].agg(sum)
print(df_res_state)

            estimated_population  estimated_population_2019  new_confirmed  \
date                                                                         
2020-03-08               3351543                    3337357              1   
2020-03-09               3351543                    3337357              0   
2020-03-10               3351543                    3337357              0   
2020-03-11               3351543                    3337357              0   
2020-03-12               3351543                    3337357              0   
...                          ...                        ...            ...   
2021-11-18                894470                     881935              6   
2021-11-19                894470                     881935              7   
2021-11-20                894470                     881935              7   
2021-11-21                894470                     881935              3   
2021-11-22                894470                     881935     

In [41]:
### Total de registros nas colunas.
df_res_city.count()

estimated_population         625
estimated_population_2019    625
new_confirmed                625
new_deaths                   625
dtype: int64

In [42]:
### Verficando se os campos possuem o mesmo valor.
pd.notnull(df_res_city[df_res_city == df_res_state]).sum()

### Total de registros em que os dados são iguais.

estimated_population         538
estimated_population_2019    538
new_confirmed                578
new_deaths                   610
dtype: int64

#### Conclusões da etapa de limpeza

- Verifiquei que nem todos os dados números batem ao somar os valores de população estimada e mortes confirmadas quando particionado por data, então é seguro assumir que devemos trabalhar a base de forma separada e mostrar que a soma dos dados de cidade não representação os dados de estado por data. 
- New deaths apresentou mais semelhança, de 610 datas tiveram mesma quantidade de mortes quando comparamos a soma de mortes no mesmo dia de cidade e estado.

### Transformação da base multas

- Criar um campo de mes ano

In [43]:
### Criando campo de mes ano
df_multas['mes_ano'] = df_multas['mes'] + '/' + df_multas['ano']

In [44]:
#### Criando coluna de valor inteiro do mês
meses_mapping = {    
    'JANEIRO': 1,
    'FEVEREIRO': 2,
    'MARÇO': 3,
    'ABRIL': 4,
    'MAIO': 5,
    'JUNHO': 6,
    'JULHO': 7,
    'AGOSTO': 8,
    'SETEMBRO': 9,
    'OUTUBRO': 10,
    'NOVEMBRO': 11,
    'DEZEMBRO': 12
}

df_multas['num_meses'] = df_multas['mes'].map(meses_mapping)


#### Criando dimensões
1. Base covid    
- Dimensão calendário
- Dimensão cidade
- Dimensão estado
       
2. Base multas

- Dimensão estado
- Dimensão calendário

#### Criando a dimensão calendário

In [45]:

def calendario(ano, mes):
    # Obter o primeiro dia do mês
    primeiro_dia = datetime(ano, mes, 1)
    
    # Obter o último dia do mês
    ultimo_dia = primeiro_dia.replace(
        day=pd.Timestamp(ano, mes, 1).days_in_month
    )
    
    # Gerar as datas do mês
    datas_mes = pd.date_range(primeiro_dia, '2022-12-31', freq='D')
    
    # Criar DataFrame com as datas do mês
    calendario_df = pd.DataFrame({'Data': datas_mes})
    
    # Adicionar coluna com o dia da semana (0 = segunda-feira, 1 = terça-feira, ..., 6 = domingo)
    calendario_df['Dia_Semana'] = calendario_df['Data'].dt.dayofweek
    
    # Adicionar coluna com o número da semana no ano
    calendario_df['Semana_Ano'] = calendario_df['Data'].dt.isocalendar().week
    
    # Adicionar coluna com o dia do ano
    calendario_df['Dia_Ano'] = calendario_df['Data'].dt.dayofyear
    
    # Adicionar coluna com o mês e ano
    calendario_df['Mes_Ano'] = calendario_df['Data'].dt.to_period('M')
    
    return calendario_df

# Exemplo: criar um calendário para janeiro de 2023
ano = 2018
mes = 1
calendario_dim = calendario(ano, mes)

# Exibir o DataFrame
print(calendario_dim)


           Data  Dia_Semana  Semana_Ano  Dia_Ano  Mes_Ano
0    2018-01-01           0           1        1  2018-01
1    2018-01-02           1           1        2  2018-01
2    2018-01-03           2           1        3  2018-01
3    2018-01-04           3           1        4  2018-01
4    2018-01-05           4           1        5  2018-01
...         ...         ...         ...      ...      ...
1821 2022-12-27           1          52      361  2022-12
1822 2022-12-28           2          52      362  2022-12
1823 2022-12-29           3          52      363  2022-12
1824 2022-12-30           4          52      364  2022-12
1825 2022-12-31           5          52      365  2022-12

[1826 rows x 5 columns]


In [46]:
### Dimensão calendario
display(calendario_dim)


Data  Dia_Semana  Semana_Ano  Dia_Ano  Mes_Ano
0    2018-01-01           0           1        1  2018-01
1    2018-01-02           1           1        2  2018-01
2    2018-01-03           2           1        3  2018-01
3    2018-01-04           3           1        4  2018-01
4    2018-01-05           4           1        5  2018-01
...         ...         ...         ...      ...      ...
1821 2022-12-27           1          52      361  2022-12
1822 2022-12-28           2          52      362  2022-12
1823 2022-12-29           3          52      363  2022-12
1824 2022-12-30           4          52      364  2022-12
1825 2022-12-31           5          52      365  2022-12

[1826 rows x 5 columns]

#### Transformando tabelas fatos e dimensão 
- Aqui eu me preocupei em criar tabelas dimensões sem duplicações e com o id identificador sendo os dados de estado correspondente as duas tabelas e a de cidade apenas a tabela de covid.
- OBS: Se novos regristros de estados ou cidades aparecerem em futuras etls o id deles podem mudar. 

In [47]:
### Dimensão cidade tabela de covid

df_city_dim = df[(df['place_type'] == 'city')][['city','state']].drop_duplicates(subset=['city','state'])

### Criando ID_cidade

df_city_dim['id_city'] = range(1, len(df_city_dim) + 1)

### Dimensão estado da tabela covid
df_state_dim_cov = df[['state']].drop_duplicates(subset=['state'])

### Dimensão estado da tabela multas
df_state_dim_mul = df_multas[['uf']].drop_duplicates(subset=['uf']).rename(columns={'uf': 'state'})

### Juntando ambas as tabelas pela string
df_state_dim = pd.concat([df_state_dim_cov, df_state_dim_mul], ignore_index=True)

# Criando id estado 
df_state_dim['id_state'] = range(1, len(df_state_dim) + 1)

print(df_state_dim)

   state  id_state
0     AC         1
1     AL         2
2     RJ         3
3     RS         4
4     ES         5
5     MG         6
6     SC         7
7     BA         8
8     MS         9
9     PR        10
10    SP        11
11    MT        12
12    GO        13
13    DF        14


In [48]:
##### Tabelas dimensão da tabela de multas

df_infracao = df_multas[['escopo_autuacao']].drop_duplicates(subset=['escopo_autuacao'])

##### Criando id
df_infracao['id_infracao'] = range(1, len(df_infracao) + 1)

##### Removendo descrição
df_multas = pd.merge(df_multas, df_infracao, how='left' , on= 'escopo_autuacao').drop(['escopo_autuacao'], axis=1) 


In [49]:
df_multas

_id        mes   ano  uf  \
0    6407641b305a576376e6f372  FEVEREIRO  2021  RJ   
1    6407641b305a576376e6f374  FEVEREIRO  2021  RS   
2    6407641b305a576376e6f38f      MARÇO  2021  RJ   
3    6407641b305a576376e6f3aa      ABRIL  2021  ES   
4    6407641b305a576376e6f3b7      ABRIL  2021  RJ   
..                        ...        ...   ...  ..   
965  6407641b305a576376e6f6bf   NOVEMBRO  2022  ES   
966  6407641b305a576376e6f6c0   NOVEMBRO  2022  GO   
967  6407641b305a576376e6f6c1   NOVEMBRO  2022  GO   
968  6407641b305a576376e6f6d4   NOVEMBRO  2022  SC   
969  6407641b305a576376e6f6ee   DEZEMBRO  2022  GO   

                                          amparo_legal  \
0    Lei 9503/97 Lei 10561/02 Res. CONTRAN nº 210/0...   
1    Lei 9503/97 Lei 10561/02 Res. CONTRAN nº 210/0...   
2    Lei 9503/97 Lei 10561/02 Res. CONTRAN nº 210/0...   
3    Lei 9503/97 Lei 10561/02 Res. CONTRAN nº 210/0...   
4    Lei 9503/97 Lei 10561/02 Res. CONTRAN nº 210/0...   
..                                                 ...   
965  Lei 9503/97 Lei 10233/01 Res. CONTRAN nº 882/2...   
966  Lei 9503/97 Lei 10233/01 Res. CONTRAN nº 882/2...   
967  Lei 9503/97 Lei 10233/01 Res. CONTRAN nº 882/2...   
968  Lei 9503/97 Lei 10233/01 Res. CONTRAN nº 882/2...   
969  Lei 9503/97 Lei 10233/01 Res. CONTRAN nº 882/2...   

                                    descricao_infracao quantidade_autos  \
0    Transitar com o veículo com excesso de peso - ...              400   
1    Transitar com o veículo com excesso de peso - ...               17   
2    Transitar com o veículo com excesso de peso - ...              406   
3    Transitar com o veículo com excesso de peso - ...              930   
4    Transitar com o veículo com excesso de peso - ...              353   
..                                                 ...              ...   
965  Transitar com o veículo com excesso de peso - ...              590   
966  Transitar com o veículo com excesso de peso - ...               72   
967  Transitar com o veículo com excesso de peso - ...              113   
968  Transitar com o veículo com excesso de peso - ...              204   
969  Transitar com o veículo com excesso de peso - ...               59   

            mes_ano  num_meses  id_infracao  
0    FEVEREIRO/2021          2            1  
1    FEVEREIRO/2021          2            1  
2        MARÇO/2021          3            1  
3        ABRIL/2021          4            1  
4        ABRIL/2021          4            1  
..              ...        ...          ...  
965   NOVEMBRO/2022         11            1  
966   NOVEMBRO/2022         11            1  
967   NOVEMBRO/2022         11            1  
968   NOVEMBRO/2022         11            1  
969   DEZEMBRO/2022         12            1  

[970 rows x 10 columns]

In [50]:
### Fato covid removendo as decrições de cidade e estado e inserindo o id.
 
df = pd.merge(df, df_city_dim[['city','id_city']], how='left' , on= 'city').drop(['city'], axis=1)

df = pd.merge(df, df_state_dim[['state','id_state']], how='left' , on= 'state').drop(['state'], axis=1) 


#### Renomeando uf

df_multas.rename(columns={'uf': 'state'}, inplace=True)

### Subistituindo a descrição pelo id
df_multas = pd.merge(df_multas, df_state_dim[['state','id_state']], how='left' , on= 'state').drop(['state'], axis=1) 

### Criando campo de data
df_multas['data'] = df_multas['num_meses'].astype(str) + '/' + '1'  + '/' + df_multas['ano'].astype(str)

### drop de num meses 

df_multas = df_multas.drop(['num_meses'], axis=1)


### Carregamento de dados

In [51]:
df.to_csv('covid_fato.csv', index=False)
df_city_dim.to_csv('city_dim.csv', index=False, encoding='ISO-8859-1')
df_state_dim.to_csv('state_dim.csv', index=False, encoding='utf-8')
calendario_dim.to_csv('calendario.csv', index=False, encoding='utf-8')
df_multas.to_csv('multas_fato.csv', index=False, encoding='utf-8')
df_infracao.to_csv('infracao_dim.csv', index=False, encoding='utf-8')


#### Fazendo carregamento de bases fatos distintas
df[(df['place_type'] == 'city')].to_csv('covid_fato_city.csv', index=False)
df[(df['place_type'] == 'state')].drop(['id_city'], axis=1).to_csv('covid_fato_state.csv', index=False)


#### Conclusão

- O output desse ETL será utilizado no Power Bi onde será criado o star-schema e as visões do desafio.
- Optei por dividir a base fato de covid em duas, uma representando os dados da cidade e outra os dados do estado como já mencionado.
- Vale ressaltar que as configurações regionais no meu Power Bi estão em inglês (USA), se a do avaliador estiver como configuração no Brasil os dados podem sofrer com erro. 
